[Storage] [DataMovement] Support for OAuth Support for Sync Service to Service Copy

amnguye opened the pull request:
Resolves https://github.com/Azure/azure-sdk-for-net/issues/35648

In order to support OAuth for Sync Service To Service Copy the following was done:
- ~Exposing a protected method to obtain the `TokenCredential` on the `BlobBaseClient`~
- ~Internally storing the `TokenCredential` on `BlobServiceClient` and `BlobContainerClient` to grab later in the case of a Service To Service where if the client is used as a parent client of a `BlobBaseClient`~
- ~Converting `TokenCredential` to a `HttpAuthorization` with a `Bearer` Scheme~

Other things that were updated:
- Service to Service Copy Sample to work correctly
- Renabled OAuth unit tests for Storage Resources

**Updated description since a lot has changed since the initial PR was put out:**
**Updated again to return a HttpAuthorization object**

In order to support OAuth for Sync Service To Service Copy the following was done:
- Exposing a protected method to obtain the ~`AccessToken`~ `HttpAuthorization` on the `BlobBaseClient` via the `GetCopyAuthorizationTokenAsync` method
- Internally storing the `TokenCredential` on `BlobServiceClient` and `BlobContainerClient` to grab later in the case of a Service To Service where if the client is used as a parent client of a `BlobBaseClient`
- Converting TokenCredential to `AccessToken` in the `BlobBaseClient` 
- Converting `AccessToken` to `HttpAuthorization` with a Bearer Scheme
 in the ServiceToServiceJobPart in the `TransferManager`
- Added `StorageResourceSingle.GetCopyAuthorizationHeaderAsync` to allow a way to grab the `HttpAuthorization`

This will allow room in the future to make some enhancement on caching the token on the TransferJob, so we're not grabbing a new token for every single blob in a container, we can just use the same one (until it expires).

amnguye committed:
Added unit test to convert TokenCredential to HttpAuthorization using respective scopes

amnguye committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net

amnguye committed:
WIP

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

amnguye committed:
Added more tests for page and append blob for oauth

amnguye committed:
WIP; removed InternalsVisibleTo since it wouldn't work without creating double dependency, changed internal to protected for GetTokenCredential

amnguye committed:
Added Support for OAuth in Service Sync Copy; Added Tests; Added protected API to obtain TokenCredentials

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Storage.Blobs](https://apiview.dev/Assemblies/Review/6120edc621ad4e9589a8e150cdf3044f)

amnguye committed:
Fix to constructor with default options bag for container client

amnguye committed:
Removed unnecessary work of exposing GetSasCredential

amnguye commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.net6.0.cs:
```diff
@@ -1594,6 +1594,7 @@ public partial class BlobBaseClient
         public virtual System.Threading.Tasks.Task<Azure.Response<Azure.Storage.Blobs.Models.BlobProperties>> GetPropertiesAsync(Azure.Storage.Blobs.Models.BlobRequestConditions conditions = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public virtual Azure.Response<Azure.Storage.Blobs.Models.GetBlobTagResult> GetTags(Azure.Storage.Blobs.Models.BlobRequestConditions conditions = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public virtual System.Threading.Tasks.Task<Azure.Response<Azure.Storage.Blobs.Models.GetBlobTagResult>> GetTagsAsync(Azure.Storage.Blobs.Models.BlobRequestConditions conditions = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        protected static Azure.Core.TokenCredential GetTokenCredential(Azure.Storage.Blobs.Specialized.BlobBaseClient client) { throw null; }
```
@schaabs @seanmcc-msft 

This actually follows the pattern that currently exists between `BlobServiceClient` and `BlobBatch` to pass the Authentication Policy to authenticate the `BlobBatchClient`
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.net6.0.cs#L173

Let me know what you think about this exposure, since I know we've done our best not to do this to pass our credentials.

amnguye commented on sdk/storage/Azure.Storage.Blobs/src/BlobContainerClient.cs:
```diff
@@ -338,22 +357,25 @@ public BlobContainerClient(Uri blobContainerUri, TokenCredential credential, Blo
-                transferValidation: options.TransferValidation,
-                encryptionScope: options.EncryptionScope,
-                trimBlobNameSlashes: options.TrimBlobNameSlashes);
+                        pipeline: options.Build(authentication),
```
TODO: remove unnecessary tabbing.

amnguye commented on sdk/storage/Azure.Storage.DataMovement.Blobs/samples/Sample01b_HelloWorldAsync.cs:
```diff
@@ -502,7 +502,7 @@ public async Task DownloadDirectory_EventHandler_ActiveDirectoryAuthAsync()
                     if (args.Exception.Message.Contains("500"))
                     {
-                        Console.WriteLine("We're getting throttled stop trying and lets try later");
+                        Console.WriteLine(args.Exception.Message);
```
NIT- change back

amnguye commented on sdk/storage/Azure.Storage.DataMovement/src/Models/StorageResourceCopyFromUriOptions.cs:
```diff
@@ -23,7 +20,8 @@ public class StorageResourceCopyFromUriOptions
         /// <summary>
-        /// Optional.  Source authentication used to access the source blob.
+        /// Optional.  Source authentication used to access the source blob. Cannot be used in conjunction with
+        /// SasCredential.
```
NIT - update to remove SasCredential

amnguye committed:
Cleanup

amnguye commented:
Talked offline with Scott, going to take a different approach where we don't expose the TokenCredential on the BlobBaseClient and attempt to pass the BlobBaseClient to the options bag so we can handle those credentials internally.

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

amnguye committed:
WIP

amnguye committed:
Change from returning TokenCredential to Http Authorization Header

amnguye committed:
Export API

amnguye committed:
Cleanup

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

amnguye committed:
Fix to mock blob resource tests and also added token credential to save in page blob and append blob storage resources

amnguye committed:
Add oauth tests for page and append blbos

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

amnguye committed:
Rename StorageExtension.GetHttpAuthorization to GetCopyAuthorizationHeader

amnguye commented on sdk/storage/Azure.Storage.Common/src/Shared/StorageExtensions.cs:
```diff
@@ -59,5 +59,18 @@ public static string GenerateBlockId(long offset)
             return Convert.ToBase64String(id);
         }
+
+        public static HttpAuthorization ToHttpAuthorization(
```
done

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

jalauzon-msft commented on sdk/storage/Azure.Storage.DataMovement/src/StorageResourceBase.cs:
```diff
@@ -36,5 +46,23 @@ public abstract class StorageResourceBase
+        // <summary>
+        // Internal constructor to accept the authorization Scheme
+        // </summary>
+        //protected void SetAuthorizationScheme(AzureSasCredential sasCredential)
```
We really shouldn't be putting SAS in the base package since that is an Azure Storage specific concept. `HttpAuthorization ` is a little more standard so I could see the argument for that one I guess.

jalauzon-msft commented on sdk/storage/Azure.Storage.DataMovement/src/ServiceToServiceJobPart.cs:
```diff
@@ -414,5 +420,19 @@ internal async Task DisposeHandlers()
             }
         }
+
+        private StorageResourceCopyFromUriOptions GetCopyFromUriOptions()
```
Why does the base package need to know about auth scheme at all? Can't the implementation of the resource just know what auth it should use and just do it rather than the base class telling it? Maybe I'm missing something here.

jalauzon-msft reviewed (commented):
Left some comments but I think I'm a little confused about the design overall. Maybe we can chat offline about it to help me better understand.

amnguye committed:
Use PublicAccessType.None instead to ensure OAuth is working

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

amnguye committed:
Merge continue

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

amnguye committed:
Replaced getting authorization header with access token for flexibility; Added StorageResource ability to grab the token

amnguye commented:
Updated description since a lot has changed since the initial PR was put out:

- Exposing a protected method to obtain the `AccessToken` on the `BlobBaseClient` via the `GetCopyAuthorizationTokenAsync` method
- Internally storing the `TokenCredential` on `BlobServiceClient` and `BlobContainerClient` to grab later in the case of a Service To Service where if the client is used as a parent client of a `BlobBaseClient`
- Converting TokenCredential to `AccessToken` in the `BlobBaseClient` 
- Converting `AccessToken` to `HttpAuthorization` with a Bearer Scheme
 in the ServiceToServiceJobPart in the `TransferManager`
- Added `StorageResourceSingle.GetCopyAuthorizationTokenAsync` to allow a way to grab the AccessToken

This will allow room in the future to make some enhancement on caching the token on the TransferJob, so we're not grabbing a new token for every single blob in a container, we can just use the same one (until it expires).

Other things that were updated:

- Service to Service Copy Sample to work correctly
- Renabled OAuth unit tests for Storage Resources
- Updated some comments

amnguye committed:
Recorded tests

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

jalauzon-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -641,39 +634,30 @@ private void SetNameFieldsIfNull()
         #region internal static accessors for Azure.Storage.DataMovement.Blobs
         /// <summary>
-        /// Get a <see cref="BlobBaseClient"/>'s <see cref="TokenCredential"/>
+        /// Get a <see cref="BlobBaseClient"/>'s <see cref="HttpAuthorization"/>
```
nit: This returns `AccessToken` now?

jalauzon-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/Constants.cs:
```diff
@@ -714,5 +714,11 @@ internal static class ServerTimeout
+
+        internal static class CopyHttpAuthorization
+        {
+            internal static readonly string[] Scopes = { "https://storage.azure.com/.default" };
```
Don't think we actually need `Scopes` anymore since we are getting a token from the Blob client?

amnguye replied:
We actually still need this, but I guess I would move it to the `DataMovementConstants` instead since it's only used by the `DataMovement` package

amnguye replied:
Oh wait I'm wrong. We just need this in general, since it's used [here](https://github.com/Azure/azure-sdk-for-net/pull/36199/files#diff-5abef2aa6e122b00cae3a7ef4d9ccfce5d74df942ba64620d8d5a4ce7430e877)

jalauzon-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/StorageExtensions.cs:
```diff
@@ -59,5 +60,14 @@ public static string GenerateBlockId(long offset)
             return Convert.ToBase64String(id);
         }
+
+        public static async Task<AccessToken> GetCopyAuthorizationTokenAsync(
```
Do we still need this?

amnguye replied:
Yup this is being used in the `BlobBaseClient.GetCopyAuthorizationTokenAsync`

jalauzon-msft commented on sdk/storage/Azure.Storage.DataMovement.Blobs/src/AppendBlobStorageResource.cs:
```diff
@@ -254,6 +256,30 @@ public override async Task<StorageResourceProperties> GetPropertiesAsync(Cancell
         }
 
+        /// <summary>
+        /// Gets the HTTP Authorization header for the storage resource if available.
```
Nit: Docstring and return description could be updated for AccessToken now. This applies to all implementations of this method in all resources.

jalauzon-msft reviewed (approved):
Just a few minor things.

amnguye committed:
PR comments

amnguye committed:
Merge branch 'main' into feature/storage/dm-oauth-copy

schaabs reviewed (changes requested):
Overall this is very close to what we want in the end. However, I think that the caching should be handled by the client acquiring tokens rather than in the DataMovement classes. Ideally this could be done by the `BearerTokenAuthenticationPolicy` which already has the complex logic for caching, refreshing and nullifying tokens. However, even if this isn't possibly at this time I think returning `HttpAuthoriztion` from `BlobBaseClient.GetCopyAuthorizationTokenAsync` will allow us to transition to this, even if we have to start without caching. I think this is preferable to spreading out the caching logic across all deriving libraries.

amnguye committed:
Replaced returning AccessToken with HttpAuthorization

amnguye commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -641,39 +634,30 @@ private void SetNameFieldsIfNull()
-        /// <returns>The BlobServiceClient's BlobClientOptions.</returns>
-        internal static BlobClientOptions GetClientOptions(BlobBaseClient client) =>
-            new BlobClientOptions(client.ClientConfiguration.Version)
+        protected static async Task<AccessToken> GetCopyAuthorizationTokenAsync(
```
Changed it back to return a `HttpAuthorization` object

amnguye commented:
- Updated description: Changed it back to return back a `HttpAuthorization`

amnguye committed:
Updated changelogs

amnguye committed:
Cleanup from merging main
